crates/core/src/factories/core_fetch.rs

```diff
@@ -276,38 +282,36 @@ impl CoreFetch {
                     .await
                 {
                     Ok(()) => {
-                        // Re-insert the fetch request into the queue.
-                        if let Err(err) = fetch_request_tx
-                            .try_send((op_id.clone(), agent_id.clone()))
-                        {
-                            tracing::warn!("could not re-insert fetch request for op {op_id} to agent {agent_id} into queue: {err}");
-                            // Remove op id/agent id from set to prevent build-up of state.
-                            state
-                                .lock()
-                                .unwrap()
-                                .ops
-                                .remove(&(op_id, agent_id));
-                        }
-                    }
-                    Err(err) => {
-                        tracing::warn!("could not send fetch request for op {op_id} to agent {agent_id}: {err}");
                         state
                             .lock()
                             .unwrap()
-                            .cool_down_list
-                            .add_agent(agent_id.clone());
-                        // Agent is unresponsive.
-                        // Remove associated op ids from set to prevent build-up of state.
+                            .back_off_list
+                            .remove_agent(&agent_id);
+                    }
+                    Err(err) => {
+                        tracing::warn!("could not send fetch request for op {op_id} to agent {agent_id}: {err}");
                         let mut lock = state.lock().unwrap();
-                        lock.ops = lock
-                            .ops
-                            .clone()
-                            .into_iter()
-                            .filter(|(_, a)| *a != agent_id)
-                            .collect();
+                        lock.back_off_list.back_off_agent(&agent_id);
+
+                        if lock
+                            .back_off_list
+                            .is_agent_at_max_back_off(&agent_id)
+                        {
+                            lock.requests
+                                .remove(&(op_id.clone(), agent_id.clone()));
+                        }
```

Comment:
I'd expect to see a remove all for the ops here and taking the agent off the back off list?

If we're done waiting, then I think what we want is to drop all ops from this source and let the loop filter the incoming channel based on the op set. 

That means if the source comes back online then it can resume fetching before the next max timeout elapses. I'm also thinking we'll end up making a connection request for each outstanding op/source pair for this source at at the max backoff interval?

Reply:
Beautiful idea! Yes, now it makes sense to remove all ops of the unresponsive agent.
The agent could be removed from the list explicitly, but any successful request to them will do that anyway.